### PR TITLE
Adds Human Readable Embargo and Lease Dates to Index View

### LINKED ISSUE #2327: Display lease and embargo information in show view
### Descriptive summary
The only way to view embargo and lease information is to do so in edit mode. We should also display this information when viewing the item individually, as well as in search results.

### PR DESCRIPTION
**Pull request by mtribone**
Fixes #2327 

Adds Solrizer for embargo and lease in catalog controller. Updates and uses helper method for human readable dates by grabbing the first value in the array.

Changes proposed in this pull request:
* Grabs the first value from the array
* Formats the date as standard
* Uses `:stored_sortable` and `human-readable-date` helper method

@projecthydra/sufia-code-reviewers

### CODE REVIEW

**Comment by mtribone** [comment]
screen shot 2016-08-10 at 2 45 44 pm

**Comment by mtribone** [comment]
screen shot 2016-08-10 at 2 45 26 pm

File: app/helpers/sufia/sufia_helper_behavior.rb [hunk 1]
```diff
@@ -104,7 +104,9 @@ def link_to_field(name, value, label = nil, facet_hash = {})
     # @see #index_field_link params
     # @return [Date]
     def human_readable_date(options)
-      Date.parse(options[:value]).to_formatted_s(:standard)
```
Comment (mjgiarlo) [inline app/helpers/sufia/sufia_helper_behavior.rb]: Are you finding that the value of `options[:value]` is sometimes an `Array` and sometimes something else? If so, what's the something else? I'm wondering what the code paths into `human_readable_date` are, and whether we should do work upstream to ensure that `options[:value]` is consistently typed.

File: lib/generators/sufia/templates/catalog_controller.rb [hunk 2]
```diff
@@ -80,6 +80,8 @@ def self.modified_field
     config.add_index_field solr_name("resource_type", :stored_searchable), label: "Resource Type", link_to_search: solr_name("resource_type", :facetable)
     config.add_index_field solr_name("file_format", :stored_searchable), label: "File Format", link_to_search: solr_name("file_format", :facetable)
     config.add_index_field solr_name("identifier", :stored_searchable), label: "Identifier", helper_method: :index_field_link, field_name: 'identifier'
+    config.add_index_field solr_name("embargo_release_date", Solrizer::Descriptor.new(:date, :stored, :indexed)), label: "Embargo release date", helper_method: :human_readable_date
```
Comment (mjgiarlo) [inline lib/generators/sufia/templates/catalog_controller.rb]: Would `solr_name("embargo_release_date", :stored_sortable,  type: :date)` work here?

File: lib/generators/sufia/templates/catalog_controller.rb [hunk 3]
```diff
@@ -80,6 +80,8 @@ def self.modified_field
...
+    config.add_index_field solr_name("lease_expiration_date", Solrizer::Descriptor.new(:date, :stored, :indexed)), label: "Lease expiration date", helper_method: :human_readable_date
```
Comment (mjgiarlo) [inline lib/generators/sufia/templates/catalog_controller.rb]: Would `solr_name("lease_expiration_date", :stored_sortable,  type: :date)` work here?

File: lib/generators/sufia/templates/catalog_controller.rb (hunk 2 above)
Comment (mtribone) [inline lib/generators/sufia/templates/catalog_controller.rb]: The field goes missing from the index

screen shot 2016-08-10 at 3 32 06 pm

File: app/helpers/sufia/sufia_helper_behavior.rb (hunk 1 above)
Comment (mtribone) [inline app/helpers/sufia/sufia_helper_behavior.rb]: I'm not sure what it is, but was getting a type error converting array to string. I didn't expect it tot be an array, since I don't see that in Solr. Help appreciated.

screen shot 2016-08-10 at 3 36 52 pm

File: lib/generators/sufia/templates/catalog_controller.rb (hunk 2 above)
Comment (mjgiarlo) [inline lib/generators/sufia/templates/catalog_controller.rb]: That is odd. I'm seeing those two invocations return the same value:

```ruby
Solrizer.solr_name("embargo_release_date", Solrizer::Descriptor.new(:date, :stored, :indexed))
# => "embargo_release_date_dtsi" 
Solrizer.solr_name("embargo_release_date", :stored_sortable, type: :date)
# => "embargo_release_date_dtsi" 
```

File: lib/generators/sufia/templates/catalog_controller.rb (hunk 3 above)
Comment (mtribone) [inline lib/generators/sufia/templates/catalog_controller.rb]: Same as above ^

**Comment by mjgiarlo** [comment]
Needs tests to cover the changes you made, too.

File: app/helpers/sufia/sufia_helper_behavior.rb (hunk 1 above)
Comment (mjgiarlo) [inline app/helpers/sufia/sufia_helper_behavior.rb]: @mtribone OK. Could you revert the change and then post the full traceback, so we can look to reproduce and investigate?

Comment (mtribone) [inline app/helpers/sufia/sufia_helper_behavior.rb]: https://gist.github.com/mtribone/4f6561601ccc70458a0e1eaf52f8f7cc

**Comment by mjgiarlo** [comment]
Thanks, @mtribone. Ping when Travis is done doing its slow, slow dance.

**Comment by mjgiarlo** [comment]
@mtribone I see the gist you posted. How are you generating that traceback? Does the test suite cause it, or are you doing in-browser testing?

**Comment by mtribone** [comment]
The gist is the output from the rails window in my terminal. I also have it in the browser (request info, local variables, instance variables). Got a command that I can run?

**Comment by mtribone** [comment]
screen shot 2016-08-11 at 12 37 43 pm

**Comment by mjgiarlo** [comment]
@mtribone give the byebug thingie :arrow_up: a try first. that's the easiest way to find out what's being passed into that function.

**Comment by mtribone** [comment]
Okay. Put the byebug in right before the Date.parse. This?

https://gist.github.com/mtribone/4f6561601ccc70458a0e1eaf52f8f7cc

**Comment by mjgiarlo** [comment]
@mtribone once you drop into byebug, type `p options` and then gist that

**Comment by mtribone** [comment]
I get the same thing twice

https://gist.github.com/mtribone/4f6561601ccc70458a0e1eaf52f8f7cc

**Comment by mjgiarlo** [comment]
@mtribone yeah, looks like the value is an array there: `:value=>["2016-08-23T00:00:00Z"]` I'm wondering why that's an array if embargo/lease dates are single-valued. Could it be because you have stale/junky data in your fcrepo/solr?

**Comment by mtribone** [comment]
@mjgiarlo Perhaps...should I `bundle exec solr_wrapper clean` and see if I get the same result?

File: lib/generators/sufia/templates/catalog_controller.rb [hunk 4]
```diff
@@ -80,6 +80,8 @@ def self.modified_field
     config.add_index_field solr_name("resource_type", :stored_searchable), label: "Resource Type", link_to_search: solr_name("resource_type", :facetable)
     config.add_index_field solr_name("file_format", :stored_searchable), label: "File Format", link_to_search: solr_name("file_format", :facetable)
     config.add_index_field solr_name("identifier", :stored_searchable), label: "Identifier", helper_method: :index_field_link, field_name: 'identifier'
+    config.add_index_field solr_name("embargo_release_date", :stored_sortable, type: :date), label: "Embargo release date", helper_method: :human_readable_date
```
Comment (mjgiarlo) [inline lib/generators/sufia/templates/catalog_controller.rb]: @jcoyne @cbeer do you know why this might pass a Solr doc to `#human_readable_date` where `:value` is an array? This is a single-valued field.

Comment (jcoyne) [inline lib/generators/sufia/templates/catalog_controller.rb]: Not off hand.

File: spec/views/catalog/_index_list_default.html.erb_spec.rb [hunk 5]
```diff
@@ -35,5 +37,9 @@
     expect(rendered).to include 'Test depositor_tesim'
     expect(rendered).to include '<span class="attribute-label h4">Rights:</span>'
     expect(rendered).to include 'Test rights_tesim'
+    expect(rendered).to include '<span class="attribute-label h4">Embargo release date:</span>'
+    expect(rendered).to include 'Test embargo_release_date_dtsi'
```
Comment (mjgiarlo) [inline spec/views/catalog/_index_list_default.html.erb_spec.rb]: if we can tweak this test to test (or include another) to ensure that dates come through human-readable, that would be ideal.

Comment (mtribone) [inline spec/views/catalog/_index_list_default.html.erb_spec.rb]: Could use some suggestions on the test for the date format. I'm thinking a view test is not ideal, but adding something for the sufia_helper_spec. Thoughts?

Comment (mjgiarlo) [inline spec/views/catalog/_index_list_default.html.erb_spec.rb]: WFM!

File: spec/helpers/sufia_helper_spec.rb [hunk 6]
```diff
@@ -264,4 +264,10 @@ def create_models(model, user1, user2)
       )).to eq("<a href=\"http://creativecommons.org/publicdomain/zero/1.0/\">CC0 1.0 Universal</a>, <a href=\"http://creativecommons.org/publicdomain/mark/1.0/\">Public Domain Mark 1.0</a>, and <a href=\"http://www.europeana.eu/portal/rights/rr-r.html\">All rights reserved</a>")
     end
   end
+
+  describe "#human_readable_date" do
+    it "ensures that the display of the date is human-readable" do
+      expect(helper.human_readable_date(value: [DateTime.new(2016, 8, 15)])).to eq("08/15/2016")
```
Comment (mjgiarlo) [inline spec/helpers/sufia_helper_spec.rb]: The build is failing because of rubocop here. Instead of calling `DateTime.new()`, why not pass in the string literal? Like:

```ruby
expect(helper.human_readable_date(value: ["2016-08-15T00:00:00Z"])).to eq("08/15/2016")
```
?

Then I'll :shipit:. Thanks!

**Comment by mjgiarlo** [comment]
Thanks, @mtribone
